SOC-274 Refactor the ConfirmEdit extension to be a Wikia extension

garthwebb opened the pull request:
This clean up work will be needed going forward to bring this code up to current coding standards and allow us to extend it more easily.

https://wikia-inc.atlassian.net/browse/SOC-274

Commit:
SOC-274 Commiting progress so far, though not working yet.

Commit:
SOC-274 Retake on the refactor.

Restarted with original CofirmEdit and lessons leanred from last attempt.  This currently works after some trivial tests.

Commit:
SOC-274 Update wfDebug to WikiaLogger

Commit:
SOC-274 Made a pass small changes:

* Update array => []
* Fix namespace issues

Commit:
SOC-274 Stylize.php run

Commit:
SOC-274 Several cleanup changes and new module:

* Moved some wgVars to be class constants
* Added FancyCaptcha module
* Fixed some wfDebug calls to be WikiaLogger

Commit:
SOC-274 Whoops, forgot to add FancyCaptcha on last commit

Commit:
SOC-274 Got FancyCaptcha working

lizlux commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,625 @@
+	}
+
+	/**
+	 * Inject whazawhoo
```
?? lol

lizlux commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,625 @@
+	/**
+	 * Inject whazawhoo
+	 *
+	 * @fixme if multiple thingies insert a header, could break
```
Let's create a ticket to have another pass at this extension for more cleanup like this. That way we can reference the ticket number in this comment, even if we don't get to it right away.

lizlux commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,625 @@
+	}
+
+	/**
+	 * Internal cache key for badlogin checks.
```
"bad login"

lizlux commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,625 @@
+		$ns = $editPage->mTitle->getNamespace();
+
+		// Special config for this NS?
+		if ( isset( $wg->CaptchaTriggersOnNamespace[$ns][$action] ) )
```
Do we have a convention in PHP to use curly braces even if the expression is only one line? I think it's best practice to include them for clarity.

lizlux commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,625 @@
+	public function confirmUserCreate( $u, &$message ) {
+		$wg = \F::app()->wg;
+		if ( $wg->CaptchaTriggers['createaccount'] ) {
+			/*	Wikia edit, fbId::47248 No one will be allowed to skip captcha for user creation.
```
We probably want to get rid of `wikia edit...` code b/c it's our extension now. Also we don't like commented out code.

lizlux commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,625 @@
+	 *
+	 * @return array of strings
+	 */
+	public function findLinks( &$editPage, $text ) {
```
Does this method need to be public? Might be worth a pass to see if all the public methods need to be. This seems like it's only being used in this class. Probably not a big deal but could cut down on ambiguity.

lizlux commented on extensions/wikia/Captcha/Captcha.setup.php:
```diff
@@ -0,0 +1,177 @@
+ * By default, sysops and registered bot accounts will be
+ * able to skip, while others have to go through it.
+ */
+$wgGroupPermissions['*'            ]['skipcaptcha'] = false;
```
I'm curious if we only show captcha to logged out users, and if so, do we need to set up permissions that affect logged in users.

lizlux commented on extensions/wikia/Captcha/Captcha.setup.php:
```diff
@@ -0,0 +1,177 @@
+	'Captcha\Module\FancyCaptcha',
+];
+
+// Always load the recaptcha shiz ... for now
```
Just for reference, from this line to the end of the file should be deleted once https://github.com/Wikia/app/pull/6116 is merged.

lizlux commented on extensions/wikia/Captcha/CaptchaController.class.php:
```diff
@@ -0,0 +1,34 @@
+	private $captcha;
+
+	public function __construct() {
+		$this->captcha = Captcha\Factory\Module::getInstance();
```
call `parent::__construct()`

garthwebb replied:
Done.

lizlux commented on extensions/wikia/Captcha/CaptchaHooks.class.php:
```diff
@@ -0,0 +1,70 @@
+	 */
+	static function getInstance() {
+		if ( !self::$handler ) {
+			self::$handler = new Handler();
```
Is it redundant to call `Captcha\Handler` since we're in the `Captcha` namespace? Just for my own info. Looks like it is (or else this is a bug).

garthwebb replied:
You have to specify it as `\Captcha\Handler()` because without the leading forward slash you'd be specifying `\Captcha\Captcha\Handler`.  If you do give the fully qualified name (`\Captcha\Handler()`) while in the `Captcha` namespace, its not wrong but PHPStorm complains that its unecessary.

lizlux replied:
word thanks

lizlux commented on extensions/wikia/Captcha/CaptchaHooks.class.php:
```diff
@@ -0,0 +1,70 @@
+		return Factory\Module::getInstance()->APIGetParamDescription( $module, $desc );
+	}
+
+	public static function onMakeGlobalVariablesScript( array &$vars ) {
```
This probably won't be needed.

lizlux commented on extensions/wikia/UserLogin/UserSignupSpecialController.class.php:
```diff
@@ -144,22 +144,13 @@ public function signupForm () {
+		$captchaObj = Captcha\Factory\Module::getInstance();
 		if( !empty( $captchaObj ) ) {
 			$this->rawHtml = $captchaObj->getForm();
 			$this->isFancyCaptcha = ( class_exists( 'FancyCaptcha' ) && $captchaObj instanceof FancyCaptcha );
```
`instanceof Captcha\Module\FancyCaptcha`

lizlux commented on extensions/wikia/UserLogin/js/UserSignup.js:
```diff
@@ -17,7 +17,7 @@
 			this.wikiaForm = new WikiaForm('#WikiaSignupForm');
 			this.submitButton = this.wikiaForm.inputs.submit;
-			this.captchaField = this.useCaptcha ? 'g-recaptcha-response' : '';
+			this.captchaField = this.useCaptcha ? window.wgCheckCaptchaField : '';
```
this will go away

lizlux commented:
This is a good first pass but I do think we should create a cleanup ticket to handle some of the other weirdness from ConfirmEdit, and reference the ticket number in some `fixme` comments.

Commit:
SOC-274 Fix code review comments

Commit:
SOC-274 Quote badlogin to highlight its a key not a misspelling

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,622 @@
+
+		// Special config for this NS?
+		if ( isset( $wg->CaptchaTriggersOnNamespace[$ns][$action] ) ) {
+			return $wg->CaptchaTriggersOnNamespace[ $ns ][ $action ];
```
No padding needed inside these brackets.

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,622 @@
+			return true;
+		}
+
+		if ( $this->captchaTriggers( $editPage, 'create' )  && !$editPage->mTitle->exists() ) {
```
You have an extra space before `&&`

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,622 @@
+	 *
+	 * @return bool true if the captcha should run
+	 */
+	public function shouldCheck( &$editPage, $newText, $section, $merged = false ) {
```
Oof, this method is still pretty nasty long. Can we add a fixme here to break this up a little bit? It's not something we have to get to now, but when we touch this file again it might inspire someone to be a hero.

garthwebb replied:
My last attempt at this refactor has this fully broken up.  I'll reapply this as a second step.  I'm being cautious because there's a lot of movement already and I don't want to make a lot of changes to something we don't fully understand.

jsutterfield replied:
Awesome :+1:

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,622 @@
+	 */
+	private function doConfirmEdit( $editPage, $newText, $section, $merged = false ) {
+		$wg = \F::app()->wg;
+		if ( $wg->Request->getVal( 'captchaid' ) ) {
```
Do you know what the difference is between `$wg->request` and `$wg->Request`? I usually see it written the former.

garthwebb replied:
They are the same.  I like to keep the casing of globals the same when I break them apart like this, e.g.:
`$wgRequest`
becomes
`$wg->Request`
Its easier to write and you don't have to change capitalization to grep for either version.

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,622 @@
+	 */
+	public function confirmUserLogin( $u, $pass, &$retval ) {
+		if ( $this->isBadLoginTriggered() ) {
+			if ( $this->isIPWhitelisted() )
```
If we're making sure to put braces around all our if statements, we should add one here as well.

jsutterfield commented on extensions/wikia/Captcha/Captcha.setup.php:
```diff
@@ -0,0 +1,177 @@
+ * @ingroup Extensions
+ */
+
+if ( !defined( 'MEDIAWIKI' ) ) {
```
Do we need to do this check here? I feel like 'MEDIAWIKI' should always be defined, and if someone is trying to run this by itself they have a whole 'nother set of problems.

garthwebb replied:
I think this is standard boilerplate on 3rd party extensions.  It's meant for:

  http://www.mediawiki.org/wiki/Register_globals#Protecting_MediaWiki_from_Register_Globals

so we can probably get rid of it

jsutterfield commented on extensions/wikia/Captcha/Captcha.setup.php:
```diff
@@ -0,0 +1,177 @@
+ * able to skip, while others have to go through it.
+ */
+$wgGroupPermissions['*'            ]['skipcaptcha'] = false;
+$wgGroupPermissions['user'         ]['skipcaptcha'] = false;
```
What's up with all the whitespace in these? If we want to line up the trues and falses, what about adding the spaces afterwards like we're doing on lines 58-64?

jsutterfield replied:
Also, since we're setting '*' to false, do we need to explicitly set 'user' and 'skipcaptcha' to false also? Won't those be covered by the wildcard?

garthwebb replied:
Its copied code, so I kept it because it was already there and does actually make it easier to read.  I'd prefer it this way if the IDE did it for us but its a pain to keep stuff like this around so I'll delete them for consistency.

As far as the settings, I think they're all here to enumerate the options.  I'll get rid of the `user` and `autoconfirmed` lines.

jsutterfield commented on extensions/wikia/Captcha/Captcha.setup.php:
```diff
@@ -0,0 +1,177 @@
+$wgHooks['AbortNewAccount'][] = 'Captcha\Hooks::confirmUserCreate';
+$wgHooks['LoginAuthenticateAudit'][] = 'Captcha\Hooks::triggerUserLogin';
+$wgHooks['UserLoginForm'][] = 'Captcha\Hooks::injectUserLogin';
+// $wgHooks['AbortLogin'][] = 'Captcha\Hooks::confirmUserLogin';
```
Can we delete this line?

garthwebb replied:
Ok.

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+		$text = wfMsg( $name );
+		# Obtain a more tailored message, if possible, otherwise, fall back to
+		# the default for edits
+		return wfEmptyMsg( $name, $text ) ? wfMsg( 'captcha-edit' ) : $text;
```
`wfMsg` -> `wfMessage()` ?

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+	 */
+	public function showHelp() {
+		$wg = \F::app()->wg;
+		$wg->Out->setPageTitle( wfMsg( 'captchahelp-title' ) );
```
`wfMsg` here, as well as lines 230 and 232 should be converted to` wfMessage`.

Commit:
SOC-274 Code review feedback

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,623 @@
+		$wg = \F::app()->wg;
+
+		if ( $wg->CaptchaWhitelistIP ) {
+			$wg = \F::app()->wg;
```
This line can be removed, looks like it was copied from 154.

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,623 @@
+
+			$this->trigger = "new account '" . $u->getName() . "'";
+			if ( !$this->captcha->passCaptcha() ) {
+				$message = wfMsg( 'captcha-createaccount-fail' );
```
Another `wfMsg` that should be converted to `wfMessage.`

jsutterfield commented on extensions/wikia/Captcha/Captcha.class.php:
```diff
@@ -0,0 +1,623 @@
+	 * @return bool whether to keep running callbacks
+	 */
+	public function triggerUserLogin( $user, $password, $retval ) {
+		$wg = \F::app()->wg;
```
You end up creating this local `$wg` variable in a lot of methods of this class. Thoughts on making it an instance variable that's initialized in the constructor?

garthwebb replied:
Good point, I should be extending WikiaObject to get this!

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+	 * @return string
+	 */
+	public function getMessage( $action ) {
+		$name = 'captcha-' . $action;
```
What do you think about adding all the other possible message keys to the docs of this method to make it easier to grep for?

garthwebb replied:
Yah, this same method in one of the child classes has those enumerated, I can add it here as well (though technically we should never get here)

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+		$params['captchaword'] = null;
+		$params['captchaid'] = null;
+
+		return true;
```
You could probably simplify this by making it
```php
	if ( $module instanceof \ApiEditPage ) {
		$params['captchaword'] = null;
		$params['captchaid'] = null;
    }

	return true;
```

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+		$desc['captchaid'] = 'CAPTCHA ID from previous request';
+		$desc['captchaword'] = 'Answer to the CAPTCHA';
+
+		return true;
```
Same simplifying logic could be used here. It's personal preference, but I find it nice to consolidate return statements if possible.

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+		if ( $info ) {
+			if ( $this->keyMatch( \F::app()->wg->Request->getVal( 'wpCaptchaWord' ), $info ) ) {
+				$this->log( "passed" );
+				$this->clearCaptcha( $info );
```
This line and line 161 can be consolidated and put onto line 156.

jsutterfield commented on extensions/wikia/Captcha/Module/BaseCaptcha.class.php:
```diff
@@ -0,0 +1,235 @@
+		$info = $this->retrieveCaptcha();
+		if ( $info ) {
+			if ( $this->keyMatch( \F::app()->wg->Request->getVal( 'wpCaptchaWord' ), $info ) ) {
+				$this->log( "passed" );
```
I noticed most these strings passed to the `log` message have a newline at the end but these here don't. Do we need to include that newline? Seems like it would be easier if we left it off so we don't have to remember to add it for all of these. It shouldn't affect how Kibana displays them, right? If we do need those newlines, then we should add them to these log calls for consistency.

jsutterfield commented on extensions/wikia/Captcha/Module/FancyCaptcha.class.php:
```diff
@@ -0,0 +1,316 @@
+				$char = $dirs[( $place + $j ) % count( $dirs )];
+				$return = $this->pickImageDir( "$directory/$char", $levels - 1 );
+				if ( $return ) {
+					return $return;
```
Can we use a different variable name here? `return $return;` doesn't tell the reader much what is being returned. Maybe `return $image;`?

jsutterfield commented on extensions/wikia/Captcha/Module/FancyCaptcha.class.php:
```diff
@@ -0,0 +1,316 @@
+	 * @return string
+	 */
+	public function getMessage( $action ) {
+		$name = 'fancycaptcha-' . $action;
```
Same question about thoughts on adding the rest of the message keys to the method docs to make them easier to search for?

Commit:
SOC-274 Inherit from WikiaObject to get access to $this->wg

Commit:
SOC-274 Updating deprecated calls to wfMsg

Commit:
SOC-274 OK, take these back out.

The broken handling of wg perms for this extension don't seem to effect $wgGroupPermissions with current settings

Commit:
SOC-274 Remove unused codepath

Commit:
SOC-274 Increased greppiness

Commit:
SOC-274 Simplifications

Commit:
SOC-274 Update var name

Commit:
Merge branch 'uconn-5' into SOC-274
